### ISSUE TITLE
[SCM-406] scm tag does not work with Subversion 1.5.1

### ISSUE THREAD

**Issue by olamy** [olamy/maven-scm#151]
**[James William Dumay](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jameswdumay)** opened **[SCM-406](https://issues.apache.org/jira/browse/SCM-406?redirect=false)** and commented

scm:checkin does not work with Subversion 1.5.1

On release:perform (which I assume calls scm:checkin) the following error occurs:

```
svn: File '/svn/private/atlassian/confluence/tags/confluence-project-2.10-m1/conf-acceptance-test/pom.xml' already exists
```

Using subversion 1.4.x is a good enough workaround.

---

**Affects:** 1.0

**Issue Links:**
- [SCM-412](https://issues.apache.org/jira/browse/SCM-412) Preparing release fails with svn client 1.5.1 and 1.5.2
 (_**"is duplicated by"**_)
- [CONTINUUM-1959](https://issues.apache.org/jira/browse/CONTINUUM-1959) Impossible to create a release with svn >= 1.5.1
 (_**"is duplicated by"**_)
- [SCM-262](https://issues.apache.org/jira/browse/SCM-262) scm:tag for subversion tagging from local version of code, not directly from repository

- [MRELEASE-427](https://issues.apache.org/jira/browse/MRELEASE-427) Add a mojo parameter for using the new remote tagging for svn scm provider (to prevent issue with svn > 1.5.0)

- [MRELEASE-375](https://issues.apache.org/jira/browse/MRELEASE-375) release plugin does not work with subversion > 1.5.0

98 votes, 64 watchers

**Comment by olamy** [comment]
**[James William Dumay](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jameswdumay)** commented

I'd like to also add that I've tested this on both Ubuntu linux and Mac OS X 10.5 using the CollabNet binaries.

**Comment by olamy** [comment]
**[James William Dumay](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jameswdumay)** commented

Changing the summary to show the correct name of operation.

**Comment by olamy** [comment]
**[Kyle](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=teqfreq)** commented

Hey guys, I downloaded the source code for the multiple dependencies (maven-release-plugin, maven-release-manager, maven-scm-providers-svn).

If the maven-release-plugin you have is configured to use the latest, it will use 1.1 version of the maven-scm-providers.  Simply changing the following line 162 in SvnTagCommand resolves this issue:

cl.createArgument().setValue( "." );
to
cl.createArgument().setValue( repository.getUrl() );

So not sure if someone would want to make an official revision for this.  For now I will deploy this to our local maven repository.

**Comment by olamy** [comment]
**[Kyle](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=teqfreq)** commented

Actually just after posting my comment I noticed this was already fixed in 1.1.1 - I guess the next step then to make this official is updating maven-release-plugin / manager?  Anyone know the plans for this?

**Comment by olamy** [comment]
**[Matthias Wessendorf](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=matzew)** commented

even wores, if your trunk is not named trunk... (b/c it is sometimes that case to have two trunks (to work on different "JSR specs")) but "trunk12_x"

Anyway, so I run the plugin like :
mvn release:prepare -DprepareRelease=true -Dresume -Dtag=http://svn.apache.org/repos/asf/myfaces/trinidad/tags/trinidad-1.2.10

now... I get this:
[ERROR] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Unable to tag SCM
Provider message:
The svn tag command failed.
Command output:
svn: Server sent unexpected return value (403 Forbidden) in response to MKACTIVITY request for '/repos/asf/!svn/act/32a097b1-8097-af46-a3c1-55150b6042bd'

[INFO] ------------------------------------------------------------------------

**Comment by olamy** [comment]
**[Matthias Wessendorf](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=matzew)** commented

for the problem above I had to do this:
http://svn.apache.org/viewvc?view=rev&revision=713040

now a "mvn release:prepare -DprepareRelease=true -Dresume" does the trick:

However, still some clean ups, to be consistent:
http://svn.apache.org/viewvc?view=rev&revision=713043
http://svn.apache.org/viewvc?view=rev&revision=713044

:-)

**Comment by olamy** [comment]
**[Kalle Korhonen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kaosko)** commented

Upgrading to svn 1.5.4 fixes the issue

**Comment by olamy** [comment]
**[George Gastaldi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gastaldi)** commented

Kalle, I upgraded to 1.5.4 and the problem was not solved.

**Comment by olamy** [comment]
**[Kalle Korhonen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kaosko)** commented

Try together with release plugin 2.0-beta-8 - though I'm now seeing some oddities, need to verify it's not a user error when I release our project the next time (in a week)

**Comment by olamy** [comment]
**[Tomislav Stojcevich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stojcevicht)** commented

svn 1.5.4 with release plugin 2.0-beta-8, the problem still exists

**Comment by olamy** [comment]
**[Kuno Baeriswyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=qnob)** commented

I'd like to confirm that the problem still exists with svn 1.5.4 and release plugin 2.0-beta-8 (on Windows). Does anybody know if  there is an offical bugreport from the subversion team I could track?

**Comment by olamy** [comment]
**[Kalle Korhonen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kaosko)** commented

Agree, still a problem in svn 1.5.4, sorry for the noise. Only one of our project succeeded a release without issues because of a funny side effect of using the buildnumber plugin - it does svn update as part of the build (if configured so); which as other have pointed out, you can do manually as well to continue release:prepare after it fails the first time.

**Comment by olamy** [comment]
**[Torsten Förtsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=torsten)** commented

Creating a copy (tag) on a working directory after performing a commit with SVN 1.5.1-1.5.4 seems to be an SVN issue
(please refer to [Error when creating tags from working copy](http://subversion.tigris.org/servlets/BrowseList?list=users&by=thread&from=686753)).

How about implementing the code change (using "cl.createArgument().setValue( repository.getUrl() );" instead of "cl.createArgument().setValue( "." ); ") suggested here by Kyle Lebel or in [SCM-262](http://jira.codehaus.org/browse/SCM-262)?

**Comment by olamy** [comment]
**[Torsten Förtsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=torsten)** commented

This would fix SCM-262

**Comment by olamy** [comment]
**[George Gastaldi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gastaldi)** commented

Torsten,

If applied that patch, you could not solve the worst scenario: others may commit on trunk, so the tag created could not really be the one that you checked out and ran tests/packaged locally.

What do you guys think about it ?

**Comment by olamy** [comment]
**[Torsten Förtsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=torsten)** commented

George,

you're right. Hard-coding the update to trunk like the patch in SCM-262 doesn't fly.

So how about the workaround suggested by Alison Winters to do "svn update" prior to the "svn copy" in ScmTagPhase?
Maybe something like

```
Index: src/main/java/org/apache/maven/shared/release/phase/ScmTagPhase.java
===================================================================
--- src/main/java/org/apache/maven/shared/release/phase/ScmTagPhase.java	(revision 720901)
+++ src/main/java/org/apache/maven/shared/release/phase/ScmTagPhase.java	(working copy)
@@ -61,8 +61,6 @@
 
         validateConfiguration( releaseDescriptor );
 
-        logInfo( relResult, "Tagging release with the label " + releaseDescriptor.getScmReleaseLabel() + "..." );
-
         ScmRepository repository;
         ScmProvider provider;
         try
@@ -80,12 +78,35 @@
             throw new ReleaseExecutionException( "Unable to configure SCM repository: " + e.getMessage(), e );
         }
 
+    	// TODO: want includes/excludes?
+        ScmFileSet fileSet = new ScmFileSet( new File( releaseDescriptor.getWorkingDirectory() ) );
+        
+        // Due to a Subversion bug http://subversion.tigris.org/servlets/BrowseList?list=users&by=thread&from=686753
+        // regarding the tag command (svn copy) we need to update the working directory first! 
+        if ("svn".equals(provider.getScmType()))
+        {
+	        UpdateScmResult result;
+	        try
+	        {
+            	logInfo( relResult, "Due to a Subversion bug we have to update the working directory before tagging ..." );
+            	result = provider.update( repository, fileSet );
+	        }
+	        catch ( ScmException e )
+	        {
+	            throw new ReleaseExecutionException( "An error is occurred while updating working directory: " + e.getMessage(), e );
+	        }
+	
+	        if ( !result.isSuccess() )
+	        {
+	            throw new ReleaseScmCommandException( "Unable to update working directory", result );
+	        }
+        }
+        
         TagScmResult result;
         try
         {
-            // TODO: want includes/excludes?
-            ScmFileSet fileSet = new ScmFileSet( new File( releaseDescriptor.getWorkingDirectory() ) );
             String tagName = releaseDescriptor.getScmReleaseLabel();
+            logInfo( relResult, "Tagging release with the label " + tagName + "..." );
             result = provider.tag( repository, fileSet, tagName,
                                    releaseDescriptor.getScmCommentPrefix() + " copy for tag " + tagName );
         }
```

**Comment by olamy** [comment]
**[Kyle](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=teqfreq)** commented

This is true.  My suggestion is flawed in a scenario where many developers risk committing to the trunk while building tags.  In our situation, there are often many projects on the go with a manageable amount of developers (2-5) working on them.  While building tags, we communicate to them to not commit to the trunk until the release is completed.

But obviously this does not cover all scenarios.

**Comment by olamy** [comment]
**[Alison Winters](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alisonatwork)** commented

I think Torsten is on the right track here.  If you svn update the directory before tagging it will work by using the current directory (".").  To protect against the scenario where someone else might have updated trunk in between i think we could look at the following fix (sorry for pseudo code, i'm not familiar enough with the Maven internals to provide a real patch).

```
boolean forceUpdate = getForceUpdateArgument();
int revision = getHeadRevision();
if (!forceUpdate && hasChanges(revision)) {
  println("Head revision has the following changes, do you wish to update?");
  printChanges(revision);
  if (!yesNo()) {
    return;
  }
}
update(revision);
doTag(".");
```

There is still a tiny race condition where someone checks something into head in between when you get the latest revision number and when you hit "yes" in non-force-update mode, but if that fails then it's no worse than the situation right now.  This will at least fix the case for most of us who have segmented trees where it is very unlikely we will be tagging something from a branch that anyone else is simultaneously editing.

**Comment by olamy** [comment]
**[Justin Edelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=justinedelson)** commented

Has anyone tested this against Subversion 1.6 RC2?

**Comment by olamy** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

So I close this issue because it's now possible to tag remotely with the svnexe provider.
Concerning issue with the release plugin, have a look at MRELEA[SE-427](https://issues.apache.org/jira/browse/SE-427).

**Comment by olamy** [comment]
**[Stefan Sperling](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stsp)** commented

The underlying problem in Subversion has been fixed in 1.6.5.
See http://subversion.tigris.org/issues/show_bug.cgi?id=3119
and the 1.6.5 entry in http://svn.collab.net/repos/svn/trunk/CHANGES

Maybe the plugin should limit use of its workaround to affected versions only?
